File: sdk/go/provider/nimail.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

// nimail.cn：简单 POST API 临时邮箱，无需认证

const nimailBaseURL = "https://www.nimail.cn"

func nimailRandomLocal(length int) string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	buf := make([]byte, length)
	for i := range buf {
		buf[i] = chars[rand.Intn(len(chars))]
	}
	return string(buf)
}

// NimailGenerate 创建 nimail.cn 临时邮箱
func NimailGenerate() (*CreatedMailbox, error) {
	client := HTTPClient()
	name := nimailRandomLocal(10)
	email := fmt.Sprintf("%s@nimail.cn", name)

	body := fmt.Sprintf("mail=%s", url.QueryEscape(email))
	req, err := http.NewRequest("POST", nimailBaseURL+"/api/applymail", strings.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", getCurrentUA())
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Origin", nimailBaseURL)
	req.Header.Set("Referer", nimailBaseURL+"/")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "nimail generate"); err != nil {
		return nil, err
	}

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data struct {
		Success string `json:"success"`
		User    string `json:"user"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, fmt.Errorf("nimail: 返回非 JSON: %s", string(raw[:min(len(raw), 120)]))
	}
	if data.Success != "true" || data.User == "" {
		return nil, fmt.Errorf("nimail: 创建邮箱失败 %s", string(raw))
	}

	return &CreatedMailbox{
		Channel: "nimail",
		Email:   data.User,
		Token:   data.User,
	}, nil
}

// NimailGetEmails 获取 nimail.cn 邮件列表
func NimailGetEmails(email string) ([]NormEmail, error) {
	client := HTTPClient()

	body := fmt.Sprintf("mail=%s&time=0", url.QueryEscape(email))
	req, err := http.NewRequest("POST", nimailBaseURL+"/api/getmails", strings.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", getCurrentUA())
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Origin", nimailBaseURL)
	req.Header.Set("Referer", nimailBaseURL+"/")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "nimail getEmails"); err != nil {
		return nil, err
	}

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data struct {
		Success string            `json:"success"`
		Mail    []json.RawMessage `json:"mail"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, err
	}
	if data.Success != "true" || len(data.Mail) == 0 {
		return []NormEmail{}, nil
	}

	return NormalizeRawMessages(data.Mail, email)
}

File: sdk/npm/src/providers/nimail.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "nimail";
const BASE_URL = "https://www.nimail.cn";

const UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36";

function randomLocal(n: number): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  let r = "";
  for (let i = 0; i < n; i++) {
    r += chars[Math.floor(Math.random() * chars.length)];
  }
  return r;
}

/**
 * 创建 nimail.cn 临时邮箱
 * POST /api/applymail → 注册邮箱（前缀自选）
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const name = randomLocal(10);
  const email = `${name}@nimail.cn`;

  const resp = await fetchWithTimeout(`${BASE_URL}/api/applymail`, {
    method: "POST",
    headers: {
      "User-Agent": UA,
      "Content-Type": "application/x-www-form-urlencoded",
      Origin: BASE_URL,
      Referer: `${BASE_URL}/`,
    },
    body: `mail=${encodeURIComponent(email)}`,
  });

  if (!resp.ok) {
    throw new Error(`nimail: 创建邮箱失败 ${resp.status}`);
  }

  const data = (await resp.json()) as NimailApplyResponse;
  if (data.success !== "true" || !data.user) {
    throw new Error(`nimail: 创建邮箱失败 ${JSON.stringify(data)}`);
  }

  return {
    channel: CHANNEL,
    email: data.user,
    token: data.user,
  };
}

/**
 * 获取 nimail.cn 邮件列表
 * POST /api/getmails → 获取当前邮箱的消息
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const addr = token || email;
  if (!addr) {
    throw new Error("nimail: 缺少邮箱地址");
  }

  const resp = await fetchWithTimeout(`${BASE_URL}/api/getmails`, {
    method: "POST",
    headers: {
      "User-Agent": UA,
      "Content-Type": "application/x-www-form-urlencoded",
      Origin: BASE_URL,
      Referer: `${BASE_URL}/`,
    },
    body: `mail=${encodeURIComponent(addr)}&time=0`,
  });

  if (!resp.ok) {
    throw new Error(`nimail: 获取邮件失败 ${resp.status}`);
  }

  const data = (await resp.json()) as NimailGetResponse;
  if (data.success !== "true" || !Array.isArray(data.mail)) {
    return [];
  }

  return data.mail.map((msg) =>
    normalizeEmail(
      {
        id: String(msg.id ?? msg.time ?? ""),
        from: msg.from || msg.sender || "",
        to: addr,
        subject: msg.subject || msg.title || "",
        text: msg.text || msg.content || "",
        html: msg.html || msg.content || "",
        date: msg.date || msg.time || "",
        isRead: false,
        attachments: [],
      },
      addr,
    ),
  );
}

type NimailApplyResponse = {
  delay?: string;
  tips?: string;
  user?: string;
  success?: string;
  time?: number;
};

type NimailGetResponse = {
  to?: string;
  mail?: NimailMessage[];
  success?: string;
  time?: number;
};

type NimailMessage = {
  id?: string | number;
  from?: string;
  sender?: string;
  subject?: string;
  title?: string;
  text?: string;
  html?: string;
  content?: string;
  date?: string;
  time?: string | number;
};
